docs/performance/optimization.rst

```diff
@@ -0,0 +1,544 @@
+
+(Reference: `Issue 13818`_)
+
+.. code:: sql
+
+   SELECT DISTINCT country FROM customers;
+
+use
+
+.. code:: sql
+
+   SELECT country FROM customers GROUP BY country;
+
+and instead of
+
+.. code:: sql
+
+   SELECT COUNT(DISTINCT a) FROM t;
+
+use
+
+.. code:: sql
+
+   SELECT COUNT(a)
+   FROM (
+           SELECT a
+           FROM t
+           GROUP BY a
+           ) tmp;
+
+.. _subqueries-instead-groups:
+
+********************************************************************
+ Use subqueries instead of GROUP BY if the groups are already known
+********************************************************************
+
+Consider the following query:
+
+.. code:: sql
+
+   SELECT customerid, SUM(order_amount) AS total
+   FROM customer_orders
+   GROUP BY customerid;
+
+This looks simple but to execute it CrateDB needs to keep the full result set in
+memory for all groups.
+
+If we already know what the groups will be we can use correlated subqueries
+instead:
+
+.. code:: sql
+
+   SELECT customerid,
+     (SELECT SUM(order_amount)
+      FROM customer_orders
+      WHERE customer_orders.customerid = customers.customerid
+     ) AS total
+   FROM customers;
+
+.. _batch-operations:
+
+******************
+ Batch operations
+******************
+
+If you need to perform lots of UPDATEs or expensive INSERTs from SELECT, instead
+of doing them all in one go, adopt a batch approach where the operations are
+done on groups of records each time.
+
+So for instance instead of doing:
+
+.. code:: sql
+
+   UPDATE mytable SET field1=field1+1;
+
+do
+
+.. code:: shell
+
+   for id in {1..100}; do
+           crash -c "UPDATE mytable SET field1=field1+1 WHERE customer_id=$id;"
+   done
+
+.. _pagination-filters:
+
+****************************************
+ Paginate on filters instead of results
+****************************************
+
+For instance instead of
+
+.. code:: sql
+
+   SELECT deviceid, AVG(field1)
+   FROM device_data
+   GROUP BY deviceid
+   LIMIT 1000 OFFSET 5000;
+
+We can do something like
+
+.. code:: sql
+
+   WITH devices AS (
+     SELECT deviceid
+     FROM devices
+     LIMIT 5 OFFSET 25
+   )
+   SELECT deviceid, AVG(field1)
+   FROM device_data
+   WHERE device_data.deviceid IN (SELECT devices.deviceid FROM devices)
+   GROUP BY deviceid;
+
+.. _staging-tables:
+
+*****************************************************************************
+ Use staging tables for intermediate results if you are doing a lot of JOINs
+*****************************************************************************
+
+If you have many CTEs or VIEWs and need to JOIN these in some cases it can be
+effective to store the intermediate results from these into dedicated tables and
+then use these tables, while there is a cost in writing to disk and reading data
+back we can benefit from indexing and from giving the optimizer more
+straightforward execution plans that it can optimize for parallel execution
+using multiple nodes in the cluster.
```

Comment:
```suggestion
If you have many CTEs or VIEWs with a need to JOIN them, it can be benefical
to query them individually, store intermediate results into dedicated tables, and
then use these tables for JOINing.

While there is a cost in writing to disk and reading data back, the whole operation
can benefit from indexing and from giving the optimizer more straightforward
execution plans, to enable it optimizing for better parallel execution using
multiple cluster nodes.
```